nemo/tron/converter/common.py

```diff
@@ -0,0 +1,346 @@
+    """
+    assert hasattr(config, "torch_dtype"), "Expected config to have attr `torch_dtype`"
+    torch_dtype = config.torch_dtype
+    if isinstance(torch_dtype, torch.dtype):
+        return torch_dtype
+    elif isinstance(torch_dtype, str):
+        return dtype_from_str(torch_dtype)
+    else:
+        raise ValueError("torch_dtype is not of type str/torch.dtype")
+
+
+class _ModelState:
+    """
+    Helper class for used for to modify state dict of a source model during model conversion.
+    """
+
+    def __init__(self, state_dict):
+        self._state_dict = state_dict
+
+    def state_dict(self):
+        # pylint: disable=C0115,C0116
+        return self._state_dict
+
+    def to(self, dtype):
+        # pylint: disable=C0115,C0116
+        for k, v in self._state_dict.items():
+            if v.dtype != dtype:
+                logger.warning(f"Converting {k} from {v.dtype} (source model) to {dtype} (target model)")
+            self._state_dict[k] = v.to(dtype)
+
+
+class BaseImporter:
+    def __init__(self, input_path: str | Path, output_path: str | Path):
+        self.input_path = Path(input_path) if isinstance(input_path, str) else input_path
+        self.output_path = Path(output_path) if isinstance(output_path, str) else output_path
+        (self.output_path / HF_ASSETS_DIR).mkdir(parents=True, exist_ok=True)
+        self._hf_config = None
+        self._tron_config = None
+        self.hf_config.save_pretrained(str(self.output_path / HF_ASSETS_DIR))
+
+    @property
+    def tokenizer(self) -> "_HuggingFaceTokenizer":
+        """Get the tokenizer for the HF model.
+
+        Returns:
+            _HuggingFaceTokenizer: Tokenizer instance initialized from the HF model's tokenizer
+        """
+
+        return _HuggingFaceTokenizer(
+            save_hf_tokenizer_assets(str(self.input_path), str(self.output_path / HF_ASSETS_DIR))
+        )
+
+    def init_tron_model(self, cfg: GPTConfig | T5Config):
+        with megatron_cpu_init_context(cfg):
+            model = cfg.configure_model(tokenizer=self.tokenizer)
+        return [model]
+
+    def init_hf_model(self):
+        raise NotImplementedError
+
+    def convert_state(self, source, target):
+        raise NotImplementedError
+
+    @property
+    def hf_config(self) -> "PretrainedConfig":
+        raise NotImplementedError
+
+    @property
+    def tron_config(self) -> GPTConfig | T5Config:
+        raise NotImplementedError
+
+    def apply(self) -> Path:
+        """Run the conversion from HF to NeMo format.
+
+        Args:
+            output_path: Path where the converted model will be saved
+
+        Returns:
+            Path: Path to the saved NeMo model
+        """
+        source = self.init_hf_model()
+
+        with temporary_distributed_context():
+            target = self.init_tron_model(self.tron_config)
+            self.convert_state(source, target[0])
+            state = GlobalState()
+            state.cfg = ConfigContainer(
```

Comment:
this will be kind of tedious to update for each sub-level config we add. if this is all to save the model state dict, should we add an explicit API for that which takes in just what's needed?

Reply:
Yeah this is a weird edge case. We can make an explicit API, but would like to do that in a follow up PR after this is merged.